src/rulesets/rest/2022-05-25/__tests__/property-rules.test.ts

```diff
@@ -867,6 +867,207 @@ describe("body properties", () => {
+              responses: {
+                "200": {
+                  description: "",
+                  content: {
+                    "application/json": {
+                      schema: {
+                        type: "object",
+                        properties: {
+                          things: {
+                            type: "array",
+                            items: {},
+                          },
+                        },
+                      },
+                    },
+                  },
+                },
+              },
+            },
+          },
+        },
+      };
+      const ruleInputs = {
+        ...TestHelpers.createRuleInputs(afterSpec, afterSpec),
+        context,
+      };
+      const results = ruleRunner.runRulesWithFacts(ruleInputs);
+      expect(results.every((result) => result.passed)).toBe(false);
+      expect(results.filter((result) => !result.passed)).toEqual(
+        expect.arrayContaining([
+          expect.objectContaining({
+            error: "type was not found array items",
+          }),
+        ]),
+      );
+    });
+    test("succeeds if items have type", () => {
+      const ruleRunner = new RuleRunner([propertyRules]);
+      const afterSpec: OpenAPIV3.Document = {
+        ...baseOpenAPI,
+        paths: {
+          "/example": {
+            get: {
+              responses: {
+                "200": {
+                  description: "",
+                  content: {
+                    "application/json": {
+                      schema: {
+                        type: "object",
+                        properties: {
+                          things: {
+                            type: "array",
+                            items: {
+                              type: string,
+                            },
+                          },
+                        },
+                      },
+                    },
+                  },
+                },
+              },
+            },
+          },
+        },
+      };
+      const ruleInputs = {
+        ...TestHelpers.createRuleInputs(afterSpec, afterSpec),
+        context,
+      };
+      const results = ruleRunner.runRulesWithFacts(ruleInputs);
+      expect(results.every((result) => result.passed)).toBe(true);
+    });
+
+    test("succeeds if items have oneOf schema", () => {
+      const ruleRunner = new RuleRunner([propertyRules]);
+      const afterSpec: OpenAPIV3.Document = {
+        ...baseOpenAPI,
+        paths: {
+          "/example": {
+            get: {
+              responses: {
+                "200": {
+                  description: "",
+                  content: {
+                    "application/json": {
+                      schema: {
+                        type: "object",
+                        properties: {
+                          things: {
+                            type: "array",
+                            items: {
+                              oneOf: [],
```

Comment:
Should we pass if it's an _empty_ composite type though? I think we should require at least one concrete type in here.

I suppose the worst pathological case would be nested composite types with nothing at the bottom -- so to be _really_ comprehensive we should do a recursive descent to make sure all "leaf-node" type-paths are defined (not empty composites). This would be the most formally-correct solution.

For example, this type contains a whole lot of choices, all leading to "no type defined":

```
items:
  oneOf:
    - oneOf: []
    - {}
    - anyOf:
        - allOf: []
        - {}
```

A helper function `isCompletelyDefinedType` could probably tackle this.

A band-aid quick-fix solution might be to check that "all composites must contain one or more simple non-empty type" and deal with the formal solution in a followup. AFAIK no one's trying to nest composites at the moment (but it may happen unintentionally later!)

Reply:
I was under the impression that there were other checks for composite types having concrete types--if not, wouldn't it be better to have those checks across all schemas rather than making that part of this check?

Reply:
There might be... I'll have to look (it's been a little while)

Reply:
Did a quick check and there aren't; I'm going to file an issue for validating that composites are completely typed, and go with a band-aid helper in this branch.

Reply:
https://github.com/snyk/sweater-comb/issues/330